### PR TITLE
Update groff-version to 1.23.0 from 1.22.4

### PR DESCRIPTION
**Pull request by zosopentoolsmain**

### CODE REVIEW

**Comment by MikeFultonDev** [comment]
code doesn't build yet - need to figure out why... clang _should_ be able to build it so something I don't understand

**Comment by MikeFultonDev** [comment]
This is the error when trying to build:
```
  CXX      src/libs/libdriver/printer.o
In file included from src/devices/grodvi/dvi.cpp:19:
In file included from ./lib/assert.h:26:
In file included from /c390/archive/zosv2r4/S2243/util/usr/include/assert.h:46:
./lib/stdio.h:959:35: error: expected ')'
                  (FILE *restrict fp, const char *restrict format, ...)
                                  ^
./lib/stdio.h:959:19: note: to match this '('
                  (FILE *restrict fp, const char *restrict format, ...)
                  ^
./lib/stdio.h:960:19: error: '__format__' attribute parameter 2 is out of bounds
                  _GL_ATTRIBUTE_FORMAT_PRINTF_STANDARD (2, 3)
                  ^                                     ~
./lib/stdio.h:149:3: note: expanded from macro '_GL_ATTRIBUTE_FORMAT_PRINTF_STANDARD'
  _GL_ATTRIBUTE_FORMAT ((_GL_ATTRIBUTE_SPEC_PRINTF_STANDARD, formatstring_parameter, first_argument))
  ^                                                          ~~~~~~~~~~~~~~~~~~~~~~
./lib/stdio.h:114:54: note: expanded from macro '_GL_ATTRIBUTE_FORMAT'
#  define _GL_ATTRIBUTE_FORMAT(spec) __attribute__ ((__format__ spec))
                                                     ^          ~~~~
./lib/stdio.h:281:44: note: expanded from macro '_GL_FUNCDECL_RPL'
  _GL_FUNCDECL_RPL_1 (rpl_##func, rettype, parameters_and_attributes)
                                           ^~~~~~~~~~~~~~~~~~~~~~~~~
./lib/stdio.h:283:33: note: expanded from macro '_GL_FUNCDECL_RPL_1'
  _GL_EXTERN_C rettype rpl_func parameters_and_attributes
                                ^~~~~~~~~~~~~~~~~~~~~~~~~
./lib/stdio.h:961:19: error: '__nonnull__' attribute parameter 2 is out of bounds
                  _GL_ARG_NONNULL ((1, 2)));
                  ^                    ~
./lib/stdio.h:532:51: note: expanded from macro '_GL_ARG_NONNULL'
#  define _GL_ARG_NONNULL(params) __attribute__ ((__nonnull__ params))
                                                  ^           ~~~~~~
./lib/stdio.h:281:44: note: expanded from macro '_GL_FUNCDECL_RPL'
  _GL_FUNCDECL_RPL_1 (rpl_##func, rettype, parameters_and_attributes)
                                           ^~~~~~~~~~~~~~~~~~~~~~~~~
./lib/stdio.h:283:33: note: expanded from macro '_GL_FUNCDECL_RPL_1'
  _GL_EXTERN_C rettype rpl_func parameters_and_attributes
                                ^~~~~~~~~~~~~~~~~~~~~~~~~
src/devices/grodvi/dvi.cpp:304:5: error: no matching function for call to 'rpl_fprintf'
    assert(a == 0);
    ^~~~~~~~~~~~~~
/c390/archive/zosv2r4/S2243/util/usr/include/assert.h:137:8: note: expanded from macro 'assert'
      (fprintf(*((__FILEP * const) (*(__gtab_filep(__stderr_i))) ), \
       ^~~~~~~
./lib/stdio.h:954:20: note: expanded from macro 'fprintf'
#   define fprintf rpl_fprintf
                   ^~~~~~~~~~~
./lib/stdio.h:958:1: note: candidate function not viable: requires single argument 'restrict', but 5 arguments were provided
_GL_FUNCDECL_RPL (fprintf, int,
^
./lib/stdio.h:281:23: note: expanded from macro '_GL_FUNCDECL_RPL'
  _GL_FUNCDECL_RPL_1 (rpl_##func, rettype, parameters_and_attributes)
                      ^
<scratch space>:42:1: note: expanded from here
rpl_fprintf
^
src/devices/grodvi/dvi.cpp:963:7: error: no matching function for call to 'rpl_fprintf'
      assert(0);
      ^~~~~~~~~
/c390/archive/zosv2r4/S2243/util/usr/include/assert.h:137:8: note: expanded from macro 'assert'
      (fprintf(*((__FILEP * const) (*(__gtab_filep(__stderr_i))) ), \
       ^~~~~~~
./lib/stdio.h:954:20: note: expanded from macro 'fprintf'
#   define fprintf rpl_fprintf
                   ^~~~~~~~~~~
./lib/stdio.h:958:1: note: candidate function not viable: requires single argument 'restrict', but 5 arguments were provided
_GL_FUNCDECL_RPL (fprintf, int,
^
./lib/stdio.h:281:23: note: expanded from macro '_GL_FUNCDECL_RPL'
  _GL_FUNCDECL_RPL_1 (rpl_##func, rettype, parameters_and_attributes)
                      ^
<scratch space>:42:1: note: expanded from here
rpl_fprintf
^
src/devices/grodvi/dvi.cpp:977:3: error: no matching function for call to 'rpl_fprintf'
  fprintf(stream,
  ^~~~~~~
./lib/stdio.h:954:20: note: expanded from macro 'fprintf'
#   define fprintf rpl_fprintf
                   ^~~~~~~~~~~
./lib/stdio.h:958:1: note: candidate function not viable: requires single argument 'restrict', but 5 arguments were provided
_GL_FUNCDECL_RPL (fprintf, int,
^
./lib/stdio.h:281:23: note: expanded from macro '_GL_FUNCDECL_RPL'
  _GL_FUNCDECL_RPL_1 (rpl_##func, rettype, parameters_and_attributes)
                      ^
<scratch space>:42:1: note: expanded from here
rpl_fprintf
^
6 errors generated.
```

**Comment by MikeFultonDev** [comment]
Digging into the expansion and focusing on the 'restrict' keyword, I see that:

```
struct { } f;
typedef struct f FILE;
extern "C" int rpl_fprintf (FILE *restrict fp, const char *restrict format, ...) __attribute__ ((__format__ (__printf__, 2, 3))) __attribute__ ((__nonnull__ (1, 2)));
```

fails to compile but:

```
struct { } f;
typedef struct f FILE;
extern "C" int rpl_fprintf (FILE *fp, const char *format, ...) __attribute__ ((__format__ (__printf__, 2, 3))) __attribute__ ((__nonnull__ (1, 2)));
```

compiles fine. 

This seems wrong when I look at: https://en.cppreference.com/w/c/language/restrict 

Hopefully someone versed in modern C++ can provide some guidance.

**Comment by MikeFultonDev** [comment]
The underlying problem is that assert.h is being included before config.h, so the `#define restrict __restrict__` from config.h is not being processed before assert.h is included. This probably is ok on other platforms, but on z/OS, assert.h pulls in stdio.h

I have asked a question in bug-groff to see what people think the right 'fix' is.

**Comment by MikeFultonDev** [comment]
I have put patches in under the assertfix directory. If/when I can promote the fix or get an alternative, we can remove the patches.

**Comment by MikeFultonDev** [comment]
@IgorTodorovskiIBM can you review please?

**Review by MikeFultonDev — APPROVED** [review]

**Comment by MikeFultonDev** [comment]
Issue: https://savannah.gnu.org/bugs/?64910 opened against groff. This is acknowledged as a problem: https://lists.gnu.org/archive/html/groff/2023-11/msg00035.html

**Review by IgorTodorovskiIBM — APPROVED** [review]
LGTM

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,6 @@
 groff-1.22.4/
+groff-1.23.0/
```
[the remark this answers is not in the document]
Comment (MikeFultonDev) [reply .gitignore:2]: Issue: https://github.com/ZOSOpenTools/meta/issues/586 opened.

**Pull request merged**